README.md
Inferno® is a distributed operating system, originally developed at Bell Labs, but now developed and maintained by Vita Nuova® as Free Software.  Applications written in Inferno's concurrent programming language, Limbo, are compiled to its portable virtual machine code (Dis), to run anywhere on a network in the portable environment that Inferno provides.  Unusually, that environment looks and acts like a complete operating system.

The use of a high-level language and virtual machine is sensible but mundane. The interesting thing is the system's representation of services and resources.  They are represented in a file-like name hiearchy.  Programs access them using only the file operations open, read/write, and close.  The 'files' may of course represent stored data, but may also be devices, network and protocol interfaces, dynamic data sources, and services.  The approach unifies and provides basic naming, structuring, and access control mechanisms for all system resources.  A single file-service protocol (the same as Plan 9's 9P) makes all those resources available for import or export throughout the network in a uniform way, independent of location. An application simply attaches the resources it needs to its own per-process name hierarchy ('name space').

The system can be used to build portable client and server applications. It makes it straightforward to build lean applications that share all manner of resources over a network, without the cruft of much of the 'Grid' software one sees.

Inferno can run 'native' on various ARM, PowerPC, SPARC and x86 platforms but also 'hosted', under an existing operating system (including FreeBSD, Irix, Linux, MacOS X, Plan 9, and Solaris), again on various processor types.

This Bitbucket project includes source for the basic applications, Inferno itself (hosted and native), all supporting software, including the native compiler suite, essential executables and supporting files.

This README would normally document whatever steps are necessary to get your application up and running.

### What is this repository for? ###

* Quick summary
* Version
* [Learn Markdown](https://bitbucket.org/tutorials/markdowndemo)

### How do I get set up? ###

* Summary of set up
* Configuration
* Dependencies
* Database configuration
* How to run tests
* Deployment instructions

### Contribution guidelines ###

* Writing tests
* Code review
* Other guidelines

### Who do I talk to? ###

* Repo owner or admin
* Other community or team contact

CHANGES
20140525
	implement & in replacement text in sed
20140524
	utflen in styx.b (probably ought to be in Sys) supporting 21-bit unicode (issue 314)
	allow slashes in anames [fixed by kristofwyzc, issue 312]
20140518
	sh(2): exec -> run [issue 311]
20140501
	add -s option to dns in cs(8) [issue 310]
	don't send nil soa to dnscache [issue 309]
20140302
	utils/cc/funct.c - remove out-of-bounds reference
	os/ipaq1110, ip/ipaux.c, os/sa1110 - add default cases to keep compiler happy
20140223
	/dis/sig: use raise not exit for errors
20140105
	incorrect check on keyname in getauthinfo [issue 306]
20130906
	remove redundant code
20130529
	appl/cmd/cpu.b: use Dial, and neither test /dev/draw/new nor bind #d (which is wrong anyway, it's #i for drawing now)
20130423
	mkfiles/mkfile-MacOSX-386: set 32-bit mode, and macosx-version-min to 10.6 [via http://debu.gs/entries/interlude-inferno-at-work]
20121210
	utils/awk/run.c - prevent free of pointer still in use [cinap]
20121205
	appl/cmd/mk/mksubdirs [issue 290 change suggested by powerman]
20120918
	remove unused/unusable debug/setdebug from appl/lib/auth9.b and replace Keyring by Crypt
20120820
	emu/port/devip.c reset headers flag in Conv
20120725
	increase KSTACK for Linux/arm
20120615
	add configuration emu/FreeBSD/emu-g [powerman]
20120509
	utils/libmach: add missing user stack top initialisers, minor changes to sync with 9
20120507
	mkfiles/mkfile-MacOSX-386 change cc to gcc
	emu/MacOSX/win.c null out drawcursor
20120424
	utils/cc/Posix.c and utils/c2l/Posix.c: correct definition of mydup for POSIX dup/dup2
	libbio/bprint.c error handling if flush or print fails
20120305
	add -I$ROOT/(Plan9|Inferno)/include to CFLAGS in mkfiles/^(mkfile-Plan9* mkfile-Inferno*)
		change specific object type to $OBJTYPE in the same
	libmath/fdlibm/fdlibm.h and libmath/dtoa.c, if USE_FPdbleword defined, use FPdbleword to crack doubles into integers
20120304
	rabin(2) added (preliminary)
20120213
	styxservers: add .error operation (preliminary)
20120116
	lib9.h: internally rename qsort as infqsort to avoid C library clash
	lib9/qsort.c: include lib9.h
	lib9/mkfile: move qsort to IMPORTFILES from COMMONFILES
20120103
	avoid re-defining isnan; use isNaN instead
20111231
	Linux/*/include/emu.h adjustments for USE_PTHREADS for non-x86 processors
20111221
	add `...` uninterpreted string literals to limbo/lex.c, appl/cmd/limbo/lex.b
20111215
	liblogfs,libnandfs: use logfsos.h (added to include) to make library source independent of os
	add fonts/vera
	add fonts/courier
20111214
	change Linux port to use pthreads instead of clone directly
	emu/Linux: rename os.c to os-clone.c; rename osp.c to os.c
	emu/Linux/mkfile: remove -lrt, add kproc-pthreads.$O
	Linux/*/include: define USE_PTHREADS
	Linux/386/include/emu.h: increase KSTACK to 32k for Linux gethost* functions
20111003
	libinterp/heap.c:/^dtype discount size of map [issue 258]
20110820
	emu/*/devfs.c, emu/port/devfs-posix.c - stubs for osdisksize, Linux implementation of osdisksize
	emu/port/devfs-posix.c - also try SOCK_DGRAM for AF_UNIX [issue 269]
20110819
	appl/cmd/mk/mk.b - use raise "fail:..." for exit status [issue 268]
20110629
	change x86 getcallerpc to avoid use of %ebp [issue 266]
	change type of getcallerpc to uintptr from ulong
20110619
	emu/port/devfs-posix.c: put major/minor st_dev into path without mapping (with luck)
	emu/port/kproc-pthreads.c: ensure pthread stack is at least PTHREAD_STACK_MIN (plus slop)
20110526
	change osyield in Windows to use SwitchToThread instead of sleep(0) [e-mail list]
20110525
	utils/[5678kqv]c/mkfile - include -I. before $CFLAGS to ensure right gc.h [issue 264]
20110516
	merge changes from vac branch of mechiel-inferno-os (googlecode)
20110426
	emu/Linux/asm-arm.S linux system call interface changed [issue 215, mechiel]
20110419
	libinterp/tk.c:/^Tk_rect shouldn't return non-canonical rectangles
	/appl/wm/mand.b handle reshape to empty rectangles
20110408
	appl/lib/json.b - have writeval produce same (correct) form as .text for Real [issue 230]
20110407
	fix embarrassingly broken Lists->delete; add Lists->find [issue 257]
	add some extra explanation to sys-dup(2) about fildes
20110404
	update Nt/386/bin
	/limbo/com.c - count constants in qualifiers, not qualifiers [issue 212, mechiel]
	/appl/cmd/limbo/com.b [also issue 212]
20110403
	secstore(2) and secstore(1) updated to implement writing [issue 256, mechiel]
20110402
	revision to devsrv.c change [issue 244] to avoid two calls to delwaiting
20110330
	call strtol not atoi for hex numbers [issue 255, mechiel]
20110329
	utils/mkdir/mkdir.c - add -p option [mechiel, issue 250]
	emu/Linux/mkfile - try searching -lpthread after -lrt for sem_* [issue 247]
	emu/port/devfs-posix.c - treat unix domain sockets as files (open/read/write/close) [npe, rsc, issue 254]
	emu/port/devfs-posix.c - use volatile correctly; slight code tidy
	emu/lib9/errstr-posix.c replace EINTR's error string by Eintr's text
20110315
	libtk/textw.c - fix mysterious control characters for editing (D and K) [issue 253, fix from mechiel]
20110225
	split emu.h from lib9.h for all platforms
	kproc in emu is now a void type as in os
20110224
	move private mem flag to Progs [issue 240, mechiel]
20110224
20110208
	emu/Linux/os.c - use -lrt's semaphores instead of signals
	emu/Linux/mkfile - add -lrt to library list, and push out separation of X11LIBS from SYSLIBS for emu-g done ages ago
20110126
	emu/MacOSX/os.c - move a free to a more sensible place
20110117
	emu/Nt/ipif.c changed to work with ipv6 interface changes
20110116
	add Plan9/arm support [provided by richard miller]
20110110	
	utils/libmach/qdb.c [issue 245] restore renaming of qdiv as qmuldiv
20110104
	emu/MacOSX/win.c detect double clicks [issue 232, jas]
	emu/MacOSX/os.c replace erendezvous by a per-kproc semaphore for osblock/osready
20101220
	changes to srv(3) to notify client when server vanishes [issue 244, thanks to powerman and roger]
20101205
	change a few types in ipif-posix.c to appease the gods

asm/lex.c
#include "asm.h"
#include "y.tab.h"
#include "../libinterp/tab.h"

Biobuf*	bin;
Biobuf*	bout;
Biobuf	output;
int	line;
int	heapid;
char	symbol[1024];
int	nerr;
char	cmap[256];
char*	file;
Desc*	dlist;
int	dcount;
int	dseg;
List*	mdata;
Sym*	module;
Link*	links;
Link*	linkt;
int	nlink;
int	listing;
int	mustcompile;
int	dontcompile;
char	*ofile;
int	dentry;
int	pcentry;

void
usage(void)
{
	fprint(2, "usage: asm [-l] file.s\n");
	exits("usage");
}

static char*
mkfile(char *file, char *oldext, char *ext)
{
	char *ofile;
	int n, n2;

	n = strlen(file);
	n2 = strlen(oldext);
	if(n >= n2 && strcmp(&file[n-n2], oldext) == 0)
		n -= n2;
	ofile = malloc(n + strlen(ext) + 1);
	memmove(ofile, file, n);
	strcpy(ofile+n, ext);
	return ofile;
}

void
main(int argc, char *argv[])
{
	int fd;
	char *p;

	ARGBEGIN{
	case 'C':
		dontcompile++;
		break;
	case 'c':
		mustcompile++;
		break;
	case 'l':
		listing++;
		break;
	default:
		usage();
	}ARGEND;

	fmtinstall('i', iconv);
	fmtinstall('a', aconv);
	kinit();
	pcentry = -1;
	dentry = -1;

	if(argc != 1)
		usage();
	file = argv[0];
	bin = Bopen(file, OREAD);
	if(bin == 0) {
		nerr++;
		print("open: %s: %r\n", file);
		exits("errors");
	}
	p = strrchr(file, '/');
	if(p == nil)
		p = file;
	else
		p++;
	ofile = mkfile(p, ".s", ".dis");
	fd = create(ofile, OWRITE, 0666);
	if(fd < 0) {
		nerr++;
		print("can't create: %s: %r\n", ofile);
		exits("errors");
	}
	Binit(&output, fd, OWRITE);
	bout = &output;
	line = 1;
	yyparse();
	Bterm(bin);
	Bterm(bout);
	close(fd);

	if(nerr != 0){
		remove(ofile);
		exits("errors");
	}

	exits(0);
}

int
opcode(Inst *i)
{
	return keywds[i->op].op;
}

int
iconv(Fmt *f)
{
	Inst *i;
	char buf[128];

	i = va_arg(f->args, Inst*);
	if(i == nil)
		return fmtstrcpy(f, "IZ");

	switch(keywds[i->op].terminal) {
	case TOKI0:
		sprint(buf, "%s", keywds[i->op].name);
		break;
	case TOKI1:
		sprint(buf, "%s\t%a", keywds[i->op].name, i->dst);
		break;
	case TOKI3:
		if(i->reg != 0) {
			char *pre = "";
			char *post = "";
			switch(i->reg->mode) {
			case AXIMM:
				pre = "$";
				break;
			case AXINF:
				post = "(fp)";
				break;
			case AXINM:
				post = "(mp)";
			 	break;
			}
			sprint(buf, "%s\t%a, %s%ld%s, %a", keywds[i->op].name, i->src, pre, i->reg->val, post, i->dst);
			break;
		}
	case TOKI2:
		sprint(buf, "%s\t%a, %a", keywds[i->op].name, i->src, i->dst);
		break;
	}

	return fmtstrcpy(f, buf);
}

int
aconv(Fmt *f)
{
	Addr *a;
	char buf[64];

	a = va_arg(f->args, Addr*);

	if(a == nil)
		return fmtstrcpy(f, "AZ");

	if(a->mode & AIND) {		
		switch(a->mode & ~AIND) {
		case AFP:
			sprint(buf, "%ld(%d(fp))", a->val, a->off);
			break;
		case AMP:
			sprint(buf, "%ld(%d(mp))", a->val, a->off);
			break;
		}
	}
	else {
		switch(a->mode) {
		case AFP:
			sprint(buf, "%ld(fp)", a->val);
			break;
		case AMP:
			sprint(buf, "%ld(mp)", a->val);
			break;
		case AIMM:
			sprint(buf, "$%ld", a->val);
			break;
		}
	}

	return fmtstrcpy(f, buf);
}

void
kinit(void)
{
	int i;
	Sym *s;
	
	for(i = 0; keywds[i].name; i++) {
		s = enter(keywds[i].name, keywds[i].terminal);
		s->value = keywds[i].op;
	}

	enter("desc", TOKHEAP);
	enter("mp", TOKSB);
	enter("fp", TOKFP);

	enter("byte", TOKDB);
	enter("word", TOKDW);
	enter("long", TOKDL);
	enter("real", TOKDF);
	enter("string", TOKDS);
	enter("var", TOKVAR);
	enter("ext", TOKEXT);
	enter("module", TOKMOD);
	enter("link", TOKLINK);
	enter("entry", TOKENTRY);
	enter("array", TOKARRAY);
	enter("indir", TOKINDIR);
	enter("apop", TOKAPOP);
	enter("ldts", TOKLDTS);
	enter("exceptions", TOKEXCS);
	enter("exception", TOKEXC);
	enter("exctab", TOKETAB);
	enter("source", TOKSRC);

	cmap['0'] = '\0'+1;
	cmap['z'] = '\0'+1;
	cmap['n'] = '\n'+1;
	cmap['r'] = '\r'+1;
	cmap['t'] = '\t'+1;
	cmap['b'] = '\b'+1;
	cmap['f'] = '\f'+1;
	cmap['a'] = '\a'+1;
	cmap['v'] = '\v'+1;
	cmap['\\'] = '\\'+1;
	cmap['"'] = '"'+1;
}

int
escchar(char c)
{
	int n;
	char buf[32];

	if(c >= '0' && c <= '9') {
		n = 1;
		buf[0] = c;
		for(;;) {
			c = Bgetc(bin);
			if(c == Eof)
				fatal("%d: <eof> in escape sequence", line);
			if(strchr("0123456789xX", c) == 0) {
				Bungetc(bin);
				break;
			}
			buf[n++] = c;
		}
		buf[n] = '\0';
		return strtol(buf, 0, 0);
	}

	n = cmap[c];
	if(n == 0)
		return c;
	return n-1;
}

static char	*strbuf;
static int	bufmax;

static void
resizebuf(void)
{
	bufmax += Strsize;
	strbuf = realloc(strbuf, bufmax);
}

void
eatstring(void)
{
	String *s;
	int esc, c, cnt;

	esc = 0;
	for(cnt = 0;;) {
		c = Bgetc(bin);
		switch(c) {
		case Eof:
			fatal("%d: <eof> in string constant", line);

		case '\n':
			line++;
			diag("newline in string constant");
			goto done;

		case '\\':
			if(esc) {
				if(cnt >= bufmax)
					resizebuf();
				strbuf[cnt++] = c;
				esc = 0;
				break;
			}
			esc = 1;
			break;

		case '"':
			if(esc == 0)
				goto done;

			/* Fall through */
		default:
			if(esc) {
				c = escchar(c);
				esc = 0;
			}
			if(cnt >= bufmax)
				resizebuf();
			strbuf[cnt++] = c;
			break;
		}
	}
done:
	if(cnt >= bufmax)
		resizebuf();
	strbuf[cnt] = '\0';
	s = malloc(sizeof(String));
	s->len = cnt+1;
	s->string = malloc(s->len);
	memmove(s->string, strbuf, s->len);
	yylval.string = s;
}

void
eatnl(void)
{
	int c;

	line++;
	for(;;) {
		c = Bgetc(bin);
		if(c == Eof)
			diag("eof in comment");
		if(c == '\n')
			return;
	}
}

int
yylex(void)
{
	int c;

loop:
	c = Bgetc(bin);
	switch(c) {
	case Eof:
		return Eof;
	case '"':
		eatstring();
		return TSTRING;
	case ' ':
	case '\t':
	case '\r':
		goto loop;
	case '\n':
		line++;
		goto loop;
	case '.':
		c = Bgetc(bin);
		Bungetc(bin);
		if(isdigit(c))
			return numsym('.');
		return '.';
	case '#':
		eatnl();
		goto loop;
	case '(':
	case ')':
	case ';':
	case ',':
	case '~':
	case '$':
	case '+':
	case '/':
	case '%':
	case '^':
	case '*':
	case '&':
	case '=':
	case '|':
	case '<':
	case '>':
	case '-':
	case ':':
		return c;
	case '\'':
		c = Bgetrune(bin);
		if(c == '\\')
			yylval.ival = escchar(Bgetc(bin));
		else
			yylval.ival = c;
		c = Bgetc(bin);
		if(c != '\'') {
			diag("missing '");
			Bungetc(bin);
		}
		return TCONST;

	default:
		return numsym(c);
	}
}

int
numsym(char first)
{
	int c;
	char *p;
	Sym *s;
	enum { Int, Hex, Frac, Expsign, Exp } state;

	symbol[0] = first;
	p = symbol;

	if(first == '.')
		state = Frac;
	else
		state = Int;

	if(isdigit(*p++) || state == Frac) {
		for(;;) {
			c = Bgetc(bin);
			if(c < 0)
				fatal("%d: <eof> eating numeric", line);

			switch(state) {
			case Int:
				if(strchr("01234567890", c))
					break;
				switch(c) {
				case 'x':
				case 'X':
					state = Hex;
					break;
				case '.':
					state = Frac;
					break;
				case 'e':
				case 'E':
					state = Expsign;
					break;
				default:
					goto done;
				}
				break;
			case Hex:
				if(strchr("01234567890abcdefABCDEF", c) == 0)
					goto done;
				break;
			case Frac:
				if(strchr("01234567890", c))
					break;
				if(c == 'e' || c == 'E')
					state = Expsign;
				else
					goto done;
				break;
			case Expsign:
				state = Exp;
				if(c == '-' || c == '+')
					break;
				/* else fall through */
			case Exp:
				if(strchr("01234567890", c) == 0)
					goto done;
				break;
			}
			*p++ = c;
		}
	done:
		Bungetc(bin);
		*p = '\0';
		switch(state) {
		default:
			yylval.ival = strtoll(symbol, 0, 0);
			return TCONST;
		case Frac:
		case Expsign:
		case Exp:
			yylval.fval = strtod(symbol, 0);
			return TFCONST;
		}
	}

	for(;;) {
		c = Bgetc(bin);
		if(c < 0)
			fatal("%d <eof> eating symbols", line);
		/* '$' and '/' can occur in fully-qualified Java class names */
		if(c != '_' && c != '.' && c != '/' && c != '$' && !isalnum(c)) {
			Bungetc(bin);
			break;
		}
		*p++ = c;
	}

	*p = '\0';

	s = enter(symbol, TID);
	switch(s->lexval) {
	default:
		yylval.sym = s;
		break;
	case TOKI0:
	case TOKI1:
	case TOKI2:
	case TOKI3:
		yylval.ival = s->value;
		break;
	}
	return s->lexval;
}

static Sym *hash[Hashsize];

Sym *
enter(char *name, int type)
{
	Sym *s;
	ulong h;
	char *p;

	s = lookup(name);
	if(s != nil)
		return s;

	h = 0;
	for(p = name; *p; p++)
		h = h*3 + *p;
	h %= Hashsize;

	s = malloc(sizeof(Sym));
	memset(s, 0, sizeof(Sym));
	s->name = strdup(name);
	s->lexval = type;
	s->hash = hash[h];
	hash[h] = s;
	return s;
}

Sym *
lookup(char *name)
{
	Sym *s;
	ulong h;
	char *p;

	h = 0;
	for(p = name; *p; p++)
		h = h*3 + *p;
	h %= Hashsize;

	for(s = hash[h]; s; s = s->hash)
		if(strcmp(name, s->name) == 0)
			return s;
	return 0;
}

void
yyerror(char *a, ...)
{
	va_list arg;
	char buf[128];

	if(strcmp(a, "syntax error") == 0) {
		yyerror("syntax error, near symbol '%s'", symbol);
		return;
	}
	va_start(arg, a);
	vseprint(buf, buf+sizeof(buf), a, arg);
	va_end(arg);
	print("%s %d: %s\n", file, line, buf);
	if(nerr++ > 10) {
		fprint(2, "%s %d: too many errors, giving up\n", file, line);
		remove(ofile);
		exits("yyerror");
	}
}

void
fatal(char *fmt, ...)
{
	va_list arg;
	char buf[512];

	va_start(arg, fmt);
	vseprint(buf, buf+sizeof(buf), fmt, arg);
	va_end(arg);
	fprint(2, "asm: %d (fatal compiler problem) %s\n", line, buf);
	exits(buf);
}

void
diag(char *fmt, ...)
{
	int srcline;
	va_list arg;
	char buf[512];

	srcline = line;
	va_start(arg, fmt);
	vseprint(buf, buf+sizeof(buf), fmt, arg);
	va_end(arg);
	fprint(2, "%s %d: %s\n", file, srcline, buf);
	if(nerr++ > 10) {
		fprint(2, "%s %d: too many errors, giving up\n", file, line);
		remove(ofile);
		exits("errors");
	}
}

Inst*
ai(int op)
{
	Inst *i;

	i = malloc(sizeof(Inst));
	memset(i, 0, sizeof(Inst));
	i->op = op;

	return i;
}

Addr*
aa(int val)
{
	Addr *a;

	a = malloc(sizeof(Addr));
	memset(a, 0, sizeof(Addr));
	a->val = val;
	if(val <= -1073741824 && val > 1073741823)
		diag("offset out of range");

	return a;
}

vlong
dtocanon(double d)
{
	int swap;
	ulong t;
	union { double d; ulong ul[2]; } u1;
	union { vlong v; ulong ul[2]; } u2;

	u2.v = 1;
	swap = u2.ul[0];
	u1.d = 1.;
	if(u1.ul[0]){
		u1.d = d;
		u2.ul[0] = u1.ul[0];
		u2.ul[1] = u1.ul[1];
	}else{
		u1.d = d;
		u2.ul[0] = u1.ul[1];
		u2.ul[1] = u1.ul[0];
	}
	if(swap){
		t = u2.ul[0];
		u2.ul[0] = u2.ul[1];
		u2.ul[1] = t;
	}
	return u2.v;
}
